contracts/v2/marketplace/TokenMarketplaceV2.sol

```diff
@@ -0,0 +1,496 @@
+
+    emit BidRejected(_tokenId, currentOwner, currentHighestBidder, currentHighestBiddersAmount);
+  }
+
+  function acceptBid(uint256 _tokenId, uint256 _acceptedAmount)
+  public
+  whenNotPaused
+  nonReentrant
+  {
+    address currentOwner = kodaAddress.ownerOf(_tokenId);
+    require(currentOwner == msg.sender, "Not token owner");
+
+    Offer storage offer = offers[_tokenId];
+
+    uint256 winningOffer = offer.offer;
+
+    // Check valid offer and offer not replaced whilst inflight
+    require(winningOffer > 0 && _acceptedAmount >= winningOffer, "Offer amount not satisfied");
+
+    address winningBidder = offer.bidder;
+
+    delete offers[_tokenId];
+
+    // Get edition no.
+    uint256 editionNumber = kodaAddress.editionOfTokenId(_tokenId);
+
+    _handleFunds(editionNumber, winningOffer, currentOwner);
+
+    kodaAddress.safeTransferFrom(msg.sender, winningBidder, _tokenId);
+
+    emit BidAccepted(_tokenId, currentOwner, winningBidder, winningOffer);
+  }
+
+  function _refundHighestBidder(uint256 _tokenId) internal {
+    // Get current highest bidder
+    address currentHighestBidder = offers[_tokenId].bidder;
+
+    if (currentHighestBidder != address(0)) {
+
+      // Get current highest bid amount
+      uint256 currentHighestBiddersAmount = offers[_tokenId].offer;
+
+      if (currentHighestBiddersAmount > 0) {
+
+        // Clear out highest bidder
+        delete offers[_tokenId];
+
+        // Refund it
+        currentHighestBidder.transfer(currentHighestBiddersAmount);
+      }
+    }
+  }
+
+  //////////////////////////
+  // User Listing Actions //
+  //////////////////////////
+
+  function listToken(uint256 _tokenId, uint256 _listingPrice)
+  public
+  whenNotPaused {
+    // Check ownership before listing
+    address tokenOwner = kodaAddress.ownerOf(_tokenId);
+    require(tokenOwner == msg.sender, "Not token owner");
+
+    // Check price over min bid
+    require(_listingPrice >= minBidAmount, "Listing price not enough");
+
+    // List the token
+    listings[_tokenId] = Listing({
+    price : _listingPrice,
+    seller : msg.sender
+    });
+
+    emit TokenListed(_tokenId, msg.sender, _listingPrice);
+  }
+
+  function delistToken(uint256 _tokenId)
+  public
+  whenNotPaused {
+
+    // check listing found
+    require(listings[_tokenId].seller != address(0), "No listing found");
+
+    // check owner is msg.sender
+    require(kodaAddress.ownerOf(_tokenId) == msg.sender, "Only the current owner can delist");
+
+    _delistToken(_tokenId);
+  }
+
+  function buyToken(uint256 _tokenId)
+  public
+  payable
+  nonReentrant
+  whenNotPaused {
+    Listing storage listing = listings[_tokenId];
+
+    // check token is listed
+    require(listing.seller != address(0), "No listing found");
+
+    // check current owner is the lister as it may have changed hands
+    address currentOwner = kodaAddress.ownerOf(_tokenId);
+    require(listing.seller == currentOwner, "Listing not valid, token owner has changed");
```

Comment:
I guess in new contracts we would delist tokens on transfer potentially?

Reply:
Ye I think we would